`gulp dist` fails because of nailgun

## What's the issue?

Ran `gulp dist` on a newly set up Macbook Pro, got this error

```
[15:28:05] Using gulpfile ~/Development/amphtml/gulpfile.js
[15:28:05] Starting 'dist'...
[15:28:05] All packages in node_modules are up to date.
[15:28:05] Running dist. Press Ctrl + C to cancel...
[15:28:05] Building version 1908062104580 of the runtime for local testing with the prod AMP config.
[15:28:05] ⤷ Use --config={canary|prod} with your gulp dist --fortesting command to specify which config to apply.
[15:28:05] Building all AMP extensions.
[15:28:05] ⤷ Use --noextensions to skip building extensions.
[15:28:05] ⤷ Use --extensions=amp-foo,amp-bar to choose which extensions to build.
[15:28:05] ⤷ Use --extensions=minimal_set to build just the extensions needed to load article.amp.html.
[15:28:05] ⤷ Use --extensions=inabox to build just the extensions that are supported in AMPHTML ads.
[15:28:05] ⤷ Use --extensions_from=examples/foo.amp.html to build extensions from example docs.
[15:28:08] Recompiled all CSS files into build/ (2.169 s)
[15:28:08] Started nailgun-server.jar on port 2115
[15:28:08] Performing multi-pass babel transforms in /private/var/folders/3w/3m_3r7vj1hx_9twv0t5r6bgm0000gn/T/5d9dca64f9f944706bd801d7e763864f
.................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

[15:28:40] Compiled all analytics vendor configs into  dist/v0/analytics-vendors/ (615 ms)
[15:28:40] Processed 3p/nameframe.max.html (731 ms)
[15:28:40] Processed 3p/recaptcha.max.html (730 ms)
[15:28:40] Processed 3p/frame.max.html (732 ms)
[15:28:40] Copied build/css/*.css to dist/*.css (1.448 s)
Compilation failed for f.js:
Aug 07, 2019 3:28:42 PM com.facebook.nailgun.NGCommunicator readChunkImpl
WARNING: Unknown chunk type: e
Aug 07, 2019 3:28:42 PM com.facebook.nailgun.NGCommunicator lambda$startBackgroundReceive$1
WARNING: Nailgun client read future raised an exception
java.io.IOException: Unknown stream type: e
        at com.facebook.nailgun.NGCommunicator.readChunkImpl(NGCommunicator.java:523)
        at com.facebook.nailgun.NGCommunicator.readChunk(NGCommunicator.java:484)
        at com.facebook.nailgun.NGCommunicator.lambda$null$0(NGCommunicator.java:207)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:835)

com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Unterminated string at line 1 column 18170296 path $[953].sourceMap
        at com.google.gson.Gson.fromJson(Gson.java:902)
        at com.google.javascript.jscomp.AbstractCommandLineRunner.parseJsonFilesFromInputStream(AbstractCommandLineRunner.java:569)
        at com.google.javascript.jscomp.AbstractCommandLineRunner.doRun(AbstractCommandLineRunner.java:1188)
        at com.google.javascript.jscomp.AbstractCommandLineRunner.run(AbstractCommandLineRunner.java:545)
        at org.ampproject.AmpCommandLineRunner.main(Unknown Source)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:567)
        at com.facebook.nailgun.NGSession.runImpl(NGSession.java:313)
        at com.facebook.nailgun.NGSession.run(NGSession.java:199)
Caused by: com.google.gson.stream.MalformedJsonException: Unterminated string at line 1 column 18170296 path $[953].sourceMap
        at com.google.gson.stream.JsonReader.syntaxError(JsonReader.java:1559)
        at com.google.gson.stream.JsonReader.nextQuotedValue(JsonReader.java:1017)
        at com.google.gson.stream.JsonReader.nextString(JsonReader.java:815)
        at com.google.gson.internal.bind.TypeAdapters$16.read(TypeAdapters.java:418)
        at com.google.gson.internal.bind.TypeAdapters$16.read(TypeAdapters.java:406)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
        at com.google.gson.Gson.fromJson(Gson.java:887)
        ... 10 more

[15:28:42] 'dist' errored after 38 s
[15:28:42] Error in plugin 'gulp-google-closure-compiler'
Compilation errors occurred
```

## Which AMP version is affected?

Latest master at the time of writing

--- a/build-system/compile/closure-compile.js
+++ b/build-system/compile/closure-compile.js
@@ -15,6 +15,7 @@
  */
 'use strict';
 
+const argv = require('minimist')(process.argv.slice(2));
 const closureCompiler = require('google-closure-compiler');
 const colors = require('ansi-colors');
 const {highlight} = require('cli-highlight');
@@ -83,9 +84,13 @@ function gulpClosureCompile(compilerOptions, nailgunPort) {
       '../../node_modules/google-closure-compiler-java/compiler.jar'
     );
   } else {
-    // On Mac OS and Linux, speed up compilation using nailgun.
+    // On Mac OS and Linux, speed up compilation using nailgun (unless the
+    // --disable_nailgun flag was passed in)
     // See https://github.com/facebook/nailgun.
-    if (process.platform == 'darwin' || process.platform == 'linux') {
+    if (
+      !argv.disable_nailgun &&
+      (process.platform == 'darwin' || process.platform == 'linux')
+    ) {
       compilerOptions = [
         '--nailgun-port',
         nailgunPort,
@@ -95,7 +100,8 @@ function gulpClosureCompile(compilerOptions, nailgunPort) {
       pluginOptions.platform = ['native']; // nailgun-runner isn't a java binary
       initOptions.extraArguments = null; // Already part of nailgun-server
     } else {
-      // For other platforms, use AMP's custom runner.jar
+      // For other platforms, or if nailgun is explicitly disabled, use AMP's
+      // custom runner.jar
       closureCompiler.compiler.JAR_PATH = require.resolve(
         `../runner/dist/${nailgunPort}/runner.jar`
       );

--- a/build-system/tasks/check-types.js
+++ b/build-system/tasks/check-types.js
@@ -136,4 +136,10 @@ module.exports = {
   checkTypes,
 };
 
+/* eslint "google-camelcase/google-camelcase": 0 */
+
 checkTypes.description = 'Check source code for JS type errors';
+checkTypes.flags = {
+  disable_nailgun:
+    "  Doesn't use nailgun to invoke closure compiler (much slower)",
+};

--- a/build-system/tasks/dist.js
+++ b/build-system/tasks/dist.js
@@ -460,4 +460,6 @@ dist.flags = {
     '  The directory closure compiler will write out to ' +
     'with --single_pass mode. The default directory is `dist`',
   full_sourcemaps: '  Includes source code content in sourcemaps',
+  disable_nailgun:
+    "  Doesn't use nailgun to invoke closure compiler (much slower)",
 };

--- a/build-system/tasks/nailgun.js
+++ b/build-system/tasks/nailgun.js
@@ -15,6 +15,7 @@
  */
 'use strict';
 
+const argv = require('minimist')(process.argv.slice(2));
 const colors = require('ansi-colors');
 const log = require('fancy-log');
 const sleep = require('sleep-promise');
@@ -74,12 +75,16 @@ function maybeReplaceDefaultCompiler() {
  * @param {boolean} detached
  */
 async function startNailgunServer(port, detached) {
+  await maybeGenerateRunner(port);
+
+  if (argv.disable_nailgun) {
+    return;
+  }
+
   nailgunRunnerReplacer = maybeReplaceDefaultCompiler();
   if (!nailgunRunnerReplacer) {
     return;
   }
-
-  await maybeGenerateRunner(port);
 
   // Start up the nailgun server after cleaning up old instances (if any)
   const customRunner = require.resolve(`../runner/dist/${port}/runner.jar`);
@@ -131,6 +136,10 @@ async function startNailgunServer(port, detached) {
  * @param {string} port
  */
 async function stopNailgunServer(port) {
+  if (argv.disable_nailgun) {
+    return;
+  }
+
   if (nailgunRunnerReplacer) {
     nailgunRunnerReplacer.restore();
   }

